src/psyclone/psyir/backend/fortran.py

```diff
@@ -0,0 +1,448 @@
+
+
+def get_dims(symbol):
+    '''Given a Symbol instance as input, return a list of strings
+    representing the symbol's array dimensions.
+
+    :param symbol: The symbol instance.
+    :type symbol: :py:class:`psyclone.psyGen.Symbol`
+
+    :returns: the Fortran representation of the symbol's dimensions as \
+    a list.
+    :rtype: list of str
+
+    :raises NotImplementedError: if the format of the dimension is not \
+    supported.
+
+    '''
+    from psyclone.psyGen import Symbol
+
+    dims = []
+    for index in symbol.shape:
+        if isinstance(index, Symbol):
+            # references another symbol
+            dims.append(index.name)
+        elif isinstance(index, int):
+            # literal constant
+            dims.append(str(index))
+        elif index is None:
+            dims.append(":")
+        else:
+            raise NotImplementedError(
+                "unsupported get_dims index '{0}'".format(str(index)))
+    return dims
+
+
+def get_kind(symbol):
+    '''Infer the expected Fortran kind value from the Symbol
+    instance. This is a temporary LFRic-specific hack which simply
+    adds a hardcoded kind value for real variables and a hardcoded
+    kind value for integer variables. To work correctly in general the
+    symbol table needs some additional information added to it, see
+    issue #375.
+
+    :param symbol: The symbol instance.
+    :type symbol: :py:class:`psyclone.psyGen.Symbol`
+
+    :returns: the Fortran kind value for the symbol instance in lower \
+    case, or None if no kind value is required.
+    :rtype: str or NoneType
+
+    '''
+    kind = None
+    if symbol.datatype == "real":
+        kind = "r_def"
+    elif symbol.datatype == "integer":
+        kind = "i_def"
+    return kind
+
+
+class FortranWriter(PSyIRVisitor):
+    '''Implements a PSyIR-to-Fortran back end for PSyIR kernel code (not
+    currently PSyIR algorithm code which has its own gen method for
+    generating Fortran).
+
+    '''
+    def get_declaration(self, symbol):
+        '''Create and return the Fortran declaration for this Symbol.
+
+        :param symbol: The symbol instance.
+        :type symbol: :py:class:`psyclone.psyGen.Symbol`
+        :returns: The Fortran declaration as a string.
+        :rtype: str
+
+        '''
+        intent = get_intent(symbol)
+        dims = get_dims(symbol)
+        kind = get_kind(symbol)
+        result = "{0}{1}".format(self._nindent, symbol.datatype)
+        if kind:
+            result += "({0})".format(kind)
+        if dims:
+            result += ", dimension({0})".format(",".join(dims))
+        if intent:
+            result += ", intent({0})".format(intent)
+        if symbol.is_constant:
+            result += ", parameter"
+        result += " :: {0}".format(symbol.name)
+        if symbol.is_constant:
+            result += " = {0}".format(symbol.constant_value)
+        result += "\n"
+        return result
+
+    def nemokern(self, node):
```

Comment:
I must admit I was thinking of just having a single `FortranWriter`. However, you are quite right that, at the moment, NEMOKernel is not part of the PSyIR, and FortranWriter should just write out a Fortran representation of the PSyIR code. It would therefore make sense to have a NemoFortranWriter that subclasses FortranWriter and deals with all of the nodes it subclasses.

Having said that, I think that NEMOKernel is really a generic concept that we could capture in the PSyIR (using a generic name) which would remove the need to subclass `FortranWriter`.

In general, I wonder whether we really need to subclass the PSyIR. It would be better if we did not as it means we can get away with generic backends.